src/core/session.cpp

```diff
@@ -26,17 +26,63 @@ void ensureConnected(session_backend * backEnd)
     }
 }
 
+// Standard logger class used by default.
+class standard_logger_impl : public logger_impl
+{
+public:
+    standard_logger_impl()
+    {
+        logStream_ = NULL;
+    }
+
+    virtual void start_query(std::string const & query)
+    {
+        if (logStream_ != NULL)
+        {
+            *logStream_ << query << '\n';
+        }
+
+        lastQuery_ = query;
+    }
+
+    virtual void set_stream(std::ostream * s)
+    {
+        logStream_ = s;
+    }
+
+    virtual std::ostream * get_stream() const
+    {
+        return logStream_;
+    }
+
+    virtual std::string get_last_query() const
+    {
+        return lastQuery_;
+    }
+
+private:
+    virtual logger_impl* do_clone() const
```

Comment:
Wouldn't it be clearer to name it `make_clone()`?

At first, I wondered if there could be just one `clone()` method though, but I guess there is need to separate cloning action from a clone construction as two distinct operations.

Reply:
It's just the convention of using a private `do_xxx()` to "really do" whatever the public `xxx()` does, after performing some extra checks on top of it. Note that this convention is already used in SOCI (e.g. a quick search finds `do_unload()` in src/core/backend-loader.cpp). In this case, `do_clone()` makes the the real clone and the public `clone()` verifies that it has actually succeeded. It might be an overkill, but OTOH it doesn't really cost much and allows to detect problems in the custom `do_xxx()` implementation sooner and signals them via an exception rather than crashes, and it's not like logger cloning is a performance-sensitive operation, so why not?

Reply:
Your explanation makes sense. Thx